File: src/lib/api/consent.ts
/**
 * Consent API client.
 *
 * CONSENT-01 status: banner config endpoint is live. The category set is
 * sourced from the server response — the `DEFAULT_CATEGORIES` constant below
 * is retained only as an SSR/test fallback (fixtures, storybook-style
 * harnesses, and the admin preview route). The real endpoint always returns
 * an authoritative list, so consumers should prefer the server response.
 *
 * CONSENT-03 status: real endpoints wired.
 *   - `recordConsent`     → POST /api/consent/record  (public, IP rate-limited)
 *   - `fetchMyConsent`    → GET  /api/consent/me      (authed; falls back to
 *                                                      the localStorage envelope
 *                                                      for anonymous sessions
 *                                                      or when the API returns
 *                                                      401)
 *   - `submitDsarRequest` → POST /api/dsar            (public)
 *
 * The TypeScript contracts below alias the generated OpenAPI shapes. When
 * CONSENT-03 lands its own schemas the remaining local interfaces become
 * aliases the same way `BannerConfig` already is.
 */

import { api, ApiError } from '$lib/api/client';
import type { components } from '$lib/api/schema';

/** Generated OpenAPI shape — single source of truth for the banner payload. */
export type BannerConfig = components['schemas']['BannerConfig'];
export type BannerCopy = components['schemas']['BannerCopy'];
export type ConsentCategoryDef = components['schemas']['ConsentCategoryDef'];

export type BannerLayout = 'bar' | 'box' | 'popup' | 'fullscreen';
export type BannerPosition = 'top' | 'bottom' | 'center' | 'bottom-start' | 'bottom-end';

/**
 * The server accepts the full action enum from the CONSENT-03 migration.
 * The UI currently emits only the four "decision" actions; `expired` and
 * `prefill` land in CONSENT-04 when session expiry + GPC prefill ship.
 */
export type ConsentAction = 'granted' | 'denied' | 'updated' | 'revoked' | 'expired' | 'prefill';

export type DsarKind = 'access' | 'delete' | 'portability' | 'rectification' | 'opt_out_sale';

export type ConsentRecordResponse = components['schemas']['ConsentRecordResponse'];
export type MyConsentResponse = components['schemas']['MyConsentResponse'];
export type DsarSubmitResponse = components['schemas']['DsarSubmitResponse'];
export type DsarRow = components['schemas']['DsarRow'];

/**
 * Local alias so existing callers of `fetchMyConsent` keep their old shape
 * (they only consume `categories` + `decidedAt`). When every call site
 * migrates to `MyConsentResponse` this interface can go away.
 */
export interface FetchMyConsentResponse {
	readonly categories: Readonly<Record<string, boolean>>;
	readonly decidedAt: string;
}

/**
 * Canonical default category set. Used as an SSR/offline fallback so
 * the banner can render before (or without) the network round-trip, and by
 * `ConsentStore` as the seed shape for its preference record.
 *
 * Keep in sync with the seed in `backend/migrations/024_consent.sql` — the
 * test in `backend/src/handlers/consent.rs::seed_copy_json_parses` is the
 * backstop if they drift.
 */
export const DEFAULT_CATEGORIES: readonly ConsentCategoryDef[] = [
	{
		key: 'necessary',
		label: 'Strictly necessary',
		description:
			'Essential for the site to function — authentication, fraud prevention, load balancing, and your preferences. Cannot be turned off.',
		required: true,
		defaultEnabled: true
	},
	{
		key: 'functional',
		label: 'Functional',
		description:
			'Remembers choices you make (language, region, saved filters) to give you a better experience.',
		required: false,
		defaultEnabled: false
	},
	{
		key: 'analytics',
		label: 'Analytics',
		description:
			'Helps us understand how visitors use the site so we can improve it. All data is aggregated and never sold.',
		required: false,
		defaultEnabled: false
	},
	{
		key: 'marketing',
		label: 'Marketing',
		description:
			'Lets us measure campaign performance and show you relevant offers on other sites.',
		required: false,
		defaultEnabled: false
	},
	{
		key: 'personalization',
		label: 'Personalization',
		description:
			'Tailors what you see — recommendations, homepage layout, trader highlights — to your interests.',
		required: false,
		defaultEnabled: false
	}
];

/** SSR/offline fallback copy. Real copy flows from the server response. */
export const DEFAULT_COPY: BannerCopy = {
	title: 'We value your privacy',
	body: 'We use cookies and similar technologies to power the site, understand usage, and — with your permission — personalize content. You can accept everything, reject non-essential categories, or choose exactly what to allow.',
	acceptAll: 'Accept all',
	rejectAll: 'Reject all',
	customize: 'Customize',
	savePreferences: 'Save preferences',
	privacyPolicyHref: '/privacy',
	privacyPolicyLabel: 'Privacy policy'
};

/** SSR/offline fallback config. Real config flows from `/api/consent/banner`. */
export const STUB_BANNER_CONFIG: BannerConfig = {
	version: 1,
	policyVersion: 1,
	layout: 'bar',
	position: 'bottom',
	locale: 'en',
	region: 'default',
	categories: [...DEFAULT_CATEGORIES],
	copy: DEFAULT_COPY,
	theme: {}
};

/**
 * Fetch the current banner config from `/api/consent/banner`.
 *
 * Returns `STUB_BANNER_CONFIG` when the endpoint is unreachable (network
 * error, 5xx) so the UI degrades to the default experience rather than
 * failing closed. Logs the underlying error at debug-level so that
 * observability still picks it up without polluting the console in the
 * common-case dev flow where the backend is offline.
 *
 * `locale` accepts any BCP-47 tag. When omitted, the caller's
 * `navigator.language` is used so the server can pick a matching banner
 * locale without a second round-trip.
 *
 * `region` is normally NOT passed by the client — the server resolves the
 * regulatory bucket from request headers (CF-IPCountry etc.). The override
 * exists for admin-preview flows where the admin wants to see the EU
 * banner without spoofing headers.
 */
export async function fetchBannerConfig(locale?: string, region?: string): Promise<BannerConfig> {
	const params = new URLSearchParams();
	const resolvedLocale =
		locale ??
		(typeof navigator !== 'undefined' && navigator.language ? navigator.language : undefined);
	if (resolvedLocale) params.set('locale', resolvedLocale);
	if (region) params.set('region', region);
	const qs = params.toString() ? `?${params.toString()}` : '';
	try {
		return await api.get<BannerConfig>(`/api/consent/banner${qs}`, { skipAuth: true });
	} catch (err) {
		if (typeof console !== 'undefined') {
			const reason = err instanceof ApiError ? `HTTP ${err.status}` : String(err);
			console.debug('[consent] fetchBannerConfig fell back to stub:', reason);
		}
		return STUB_BANNER_CONFIG;
	}
}

/**
 * Anonymous-id cookie used for linking consent decisions across sessions
 * before the subject authenticates. 1y SameSite=Lax per ICO guidance —
 * "strictly necessary" treatment so the banner isn't gated on itself.
 */
const ANON_ID_KEY = 'swings_consent_anon_v1';

function readOrCreateAnonId(): string | undefined {
	if (typeof window === 'undefined') return undefined;
	try {
		const existing = window.localStorage.getItem(ANON_ID_KEY);
		if (existing && /^[0-9a-fA-F-]{32,36}$/.test(existing)) return existing;
		const fresh = generateUuid();
		window.localStorage.setItem(ANON_ID_KEY, fresh);
		return fresh;
	} catch {
		return undefined;
	}
}

/**
 * Record a consent event.
 *
 * The server enriches the row with `subject_id` (from the JWT if present),
 * `ip_hash`, `user_agent`, `country`, `banner_version`, and `policy_version`.
 * The client supplies the categories payload, any per-service overrides, the
 * GPC signal (if `navigator.globalPrivacyControl === true`), and an
 * anonymous-id UUID for unauthenticated subjects.
 */
export async function recordConsent(
	action: ConsentAction,
	categories: Readonly<Record<string, boolean>>,
	extras?: {
		readonly services?: Readonly<Record<string, boolean>>;
		readonly tcfString?: string;
		readonly gpcSignal?: boolean;
		readonly bannerVersion?: number;
		readonly policyVersion?: number;
	}
): Promise<ConsentRecordResponse> {
	const anonymousId = readOrCreateAnonId();
	const body: Record<string, unknown> = {
		action,
		categories
	};
	if (extras?.services) body.services = extras.services;
	if (extras?.tcfString) body.tcfString = extras.tcfString;
	if (typeof extras?.gpcSignal === 'boolean') body.gpcSignal = extras.gpcSignal;
	if (typeof extras?.bannerVersion === 'number') body.bannerVersion = extras.bannerVersion;
	if (typeof extras?.policyVersion === 'number') body.policyVersion = extras.policyVersion;
	if (anonymousId) body.anonymousId = anonymousId;
	return api.post<ConsentRecordResponse>('/api/consent/record', body, { skipAuth: false });
}

/**
 * Fetch the current subject's consent state.
 *
 * Authenticated users: calls `GET /api/consent/me` and returns the newest
 * row's `categories` + `decidedAt`. Anonymous users (401 response) fall back
 * to the localStorage envelope persisted by the consent store so the UI
 * still sees a consistent view.
 */
export async function fetchMyConsent(): Promise<FetchMyConsentResponse | null> {
	try {
		const resp = await api.get<MyConsentResponse>('/api/consent/me');
		if (!resp.decidedAt) return readLocalConsent();
		return {
			categories: (resp.categories ?? {}) as Record<string, boolean>,
			decidedAt: resp.decidedAt
		};
	} catch (err) {
		if (err instanceof ApiError && err.status === 401) {
			return readLocalConsent();
		}
		if (typeof console !== 'undefined') {
			const reason = err instanceof ApiError ? `HTTP ${err.status}` : String(err);
			console.debug('[consent] fetchMyConsent fell back to localStorage:', reason);
		}
		return readLocalConsent();
	}
}

/**
 * Submit a Data Subject Access Request.
 *
 * Returns the DSAR id the backend minted. The subject will receive a
 * verification e-mail; actual fulfilment is an admin-side workflow.
 */
export async function submitDsarRequest(
	email: string,
	kind: DsarKind,
	payload?: Readonly<Record<string, unknown>>
): Promise<DsarSubmitResponse> {
	return api.post<DsarSubmitResponse>('/api/dsar', { email, kind, payload }, { skipAuth: true });
}

function readLocalConsent(): FetchMyConsentResponse | null {
	if (typeof window === 'undefined') return null;
	try {
		const raw = window.localStorage.getItem('swings_consent_v1');
		if (!raw) return null;
		const parsed = JSON.parse(raw) as unknown;
		if (!parsed || typeof parsed !== 'object') return null;
		const envelope = parsed as Partial<{
			categories: Record<string, boolean>;
			decidedAt: string;
		}>;
		if (!envelope.categories || !envelope.decidedAt) return null;
		return { categories: envelope.categories, decidedAt: envelope.decidedAt };
	} catch {
		return null;
	}
}

function generateUuid(): string {
	if (typeof crypto !== 'undefined' && typeof crypto.randomUUID === 'function') {
		return crypto.randomUUID();
	}
	return `consent_${Math.random().toString(36).slice(2)}_${Date.now().toString(36)}`;
}

File: backend/src/handlers/consent.rs
//! CONSENT-01 — public banner config endpoint.
//!
//! `GET /api/consent/banner?locale=<bcp47>` resolves the active banner config
//! for the request's `(region, locale)` pair via
//! [`crate::consent::repo::resolve_banner`] and returns it alongside the
//! category catalogue and current policy version.
//!
//! Region resolution is a CONSENT-05 concern; this handler currently passes
//! `"default"` so the single seeded row is always returned. When CONSENT-05
//! lands the geo resolver, only [`resolve_region`] below changes — the rest
//! of the response shape is already region-aware.
//!
//! The response shape is the wire contract consumed by `src/lib/api/consent.ts`
//! on the frontend; the Svelte banner hydrates directly from this payload.
//! Admin CRUD for banners/categories/services/policies belongs to CONSENT-07
//! and lives under `/api/admin/consent/*`.

use axum::{extract::Query, extract::State, routing::get, Json, Router};
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

use crate::{
    consent::repo,
    error::{AppError, AppResult},
    AppState,
};

// ── Router ──────────────────────────────────────────────────────────────

pub fn public_router() -> Router<AppState> {
    Router::new().route("/banner", get(get_banner))
}

// ── Query + response shapes ─────────────────────────────────────────────

#[derive(Debug, Deserialize)]
pub struct BannerQuery {
    /// BCP-47 locale tag. Falls back to `en` when missing or unrecognised.
    pub locale: Option<String>,
}

/// Single-category entry in the banner response.
///
/// Wire fields are camelCased to match the frontend stub at
/// `src/lib/api/consent.ts`; when schema codegen replaces the stub the
/// Svelte components consume this shape unchanged.
#[derive(Debug, Clone, Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct ConsentCategoryDef {
    /// Stable key — MUST NOT be renamed after a row has been written to the
    /// consent log (CONSENT-03). Migration-level change only.
    pub key: String,
    pub label: String,
    pub description: String,
    /// When true, toggle is disabled in the preferences modal.
    pub required: bool,
    /// Whether the category is pre-checked before the user interacts.
    /// GDPR Art. 4(11) + EDPB 05/2020 §86: non-required categories MUST default
    /// to `false`. Derived here rather than stored so a data fix is cheap.
    pub default_enabled: bool,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "lowercase")]
pub enum BannerLayout {
    Bar,
    Box,
    Popup,
    Fullscreen,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "kebab-case")]
pub enum BannerPosition {
    Top,
    Bottom,
    Center,
    BottomStart,
    BottomEnd,
}

impl BannerLayout {
    fn parse(s: &str) -> Self {
        match s {
            "box" => Self::Box,
            "popup" => Self::Popup,
            "fullscreen" => Self::Fullscreen,
            _ => Self::Bar,
        }
    }
}

impl BannerPosition {
    fn parse(s: &str) -> Self {
        match s {
            "top" => Self::Top,
            "center" => Self::Center,
            "bottom-start" => Self::BottomStart,
            "bottom-end" => Self::BottomEnd,
            _ => Self::Bottom,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct BannerCopy {
    pub title: String,
    pub body: String,
    pub accept_all: String,
    pub reject_all: String,
    pub customize: String,
    pub save_preferences: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub privacy_policy_href: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub privacy_policy_label: Option<String>,
}

#[derive(Debug, Clone, Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct BannerConfig {
    /// Banner config row version. Bumped when copy/layout changes; used by
    /// CONSENT-03 to re-prompt subjects whose recorded consent predates this.
    pub version: i32,
    /// Current privacy-policy version, sourced from `consent_policies`.
    pub policy_version: i32,
    pub layout: BannerLayout,
    pub position: BannerPosition,
    pub locale: String,
    pub region: String,
    pub categories: Vec<ConsentCategoryDef>,
    pub copy: BannerCopy,
    /// Opaque theme overrides; the frontend maps known keys to PE7 CSS vars.
    pub theme: serde_json::Value,
}

// ── Handlers ────────────────────────────────────────────────────────────

/// Resolve the active banner config + category list + current policy version.
///
/// Always returns 200 with the `default`/`en` seed if no better match exists.
/// The only error shape is a `503` if the tables are empty (which can only
/// happen if migration `024_consent.sql` has not run).
#[utoipa::path(
    get,
    path = "/api/consent/banner",
    tag = "consent",
    params(
        ("locale" = Option<String>, Query, description = "BCP-47 locale tag; defaults to 'en'.")
    ),
    responses(
        (status = 200, description = "Resolved banner config", body = BannerConfig),
        (status = 503, description = "Consent tables not seeded")
    )
)]
pub async fn get_banner(
    State(state): State<AppState>,
    Query(q): Query<BannerQuery>,
) -> AppResult<Json<BannerConfig>> {
    let locale = normalise_locale(q.locale.as_deref());
    let region = resolve_region();

    let banner = repo::resolve_banner(&state.db, &region, &locale)
        .await?
        .ok_or_else(|| {
            AppError::ServiceUnavailable(
                "consent banner configs not seeded; run migrations".to_string(),
            )
        })?;

    let categories_rows = repo::list_categories(&state.db).await?;
    let categories = categories_rows
        .into_iter()
        .map(|c| ConsentCategoryDef {
            key: c.key,
            label: c.label,
            description: c.description,
            // EDPB 05/2020 §86 — non-required categories default to off.
            default_enabled: c.is_required,
            required: c.is_required,
        })
        .collect();

    let copy = parse_copy(&banner.copy_json).unwrap_or_else(default_copy);

    let policy = repo::latest_policy(&state.db, &banner.locale).await?;
    let policy_version = policy.map(|p| p.version).unwrap_or(1);

    Ok(Json(BannerConfig {
        version: banner.version,
        policy_version,
        layout: BannerLayout::parse(&banner.layout),
        position: BannerPosition::parse(&banner.position),
        locale: banner.locale,
        region: banner.region,
        categories,
        copy,
        theme: banner.theme_json,
    }))
}

// ── Helpers ─────────────────────────────────────────────────────────────

/// Reduce an arbitrary BCP-47 tag to the canonical lowercase primary language
/// subtag (the only form the seed rows key on in CONSENT-01). CONSENT-06 will
/// widen this to full region-aware matching once the translation catalogues
/// ship. An empty / unparseable input falls back to `en`.
fn normalise_locale(raw: Option<&str>) -> String {
    let s = raw.unwrap_or("").trim();
    if s.is_empty() {
        return "en".to_string();
    }
    let lower = s.to_ascii_lowercase();
    let primary = lower.split(['-', '_']).next().unwrap_or("en");
    if primary.is_empty() {
        "en".to_string()
    } else {
        primary.to_string()
    }
}

/// Region resolver. CONSENT-01 returns the literal `"default"`; CONSENT-05
/// will replace this with a `common::geo`-backed lookup that maps
/// country-codes onto the `EU` / `US-CA` / etc. buckets the admin configures.
fn resolve_region() -> String {
    "default".to_string()
}

fn parse_copy(value: &serde_json::Value) -> Option<BannerCopy> {
    serde_json::from_value::<BannerCopy>(value.clone()).ok()
}

/// Defaults mirror the seed in `024_consent.sql` — used as a safety net if
/// an admin ever writes an incomplete `copy_json` blob via CONSENT-07.
fn default_copy() -> BannerCopy {
    BannerCopy {
        title: "We value your privacy".to_string(),
        body: "We use cookies and similar technologies to power the site, understand usage, and — with your permission — personalize content.".to_string(),
        accept_all: "Accept all".to_string(),
        reject_all: "Reject all".to_string(),
        customize: "Customize".to_string(),
        save_preferences: "Save preferences".to_string(),
        privacy_policy_href: Some("/privacy".to_string()),
        privacy_policy_label: Some("Privacy policy".to_string()),
    }
}

// ── Unit tests ──────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn normalise_locale_defaults_to_en() {
        assert_eq!(normalise_locale(None), "en");
        assert_eq!(normalise_locale(Some("")), "en");
        assert_eq!(normalise_locale(Some("   ")), "en");
    }

    #[test]
    fn normalise_locale_extracts_primary_subtag() {
        assert_eq!(normalise_locale(Some("EN")), "en");
        assert_eq!(normalise_locale(Some("en-GB")), "en");
        assert_eq!(normalise_locale(Some("pt_BR")), "pt");
        assert_eq!(normalise_locale(Some("zh-Hant-TW")), "zh");
    }

    #[test]
    fn default_copy_parses_round_trip() {
        let dc = default_copy();
        let v = serde_json::to_value(&dc).expect("serialize");
        let back = parse_copy(&v).expect("parse");
        assert_eq!(back.title, dc.title);
        assert_eq!(back.accept_all, dc.accept_all);
        assert_eq!(back.privacy_policy_href, dc.privacy_policy_href);
    }

    #[test]
    fn parse_copy_returns_none_for_bad_shape() {
        let v = serde_json::json!({ "title": "ok" });
        assert!(parse_copy(&v).is_none(), "missing fields should fail parse");
    }

    #[test]
    fn seed_copy_json_parses() {
        // Mirrors the jsonb_build_object in `024_consent.sql` seed. If this
        // test breaks after a migration edit, the seed and the handler have
        // diverged.
        let v = serde_json::json!({
            "title": "We value your privacy",
            "body": "copy",
            "acceptAll": "Accept all",
            "rejectAll": "Reject all",
            "customize": "Customize",
            "savePreferences": "Save preferences",
            "privacyPolicyHref": "/privacy",
            "privacyPolicyLabel": "Privacy policy"
        });
        let parsed = parse_copy(&v).expect("seed shape");
        assert_eq!(parsed.accept_all, "Accept all");
        assert_eq!(parsed.save_preferences, "Save preferences");
    }
}
